torch/csrc/autograd/record_function.cpp

```diff
@@ -189,14 +196,42 @@ void RecordFunction::processCallbacks() {
 }
 
 RecordFunction::~RecordFunction() {
+  end();
+}
+
+void RecordFunction::end() {
   if (initialized_) {
     for (size_t idx = 0; idx < manager().end_callbacks.size(); ++idx) {
       if (!manager().is_callback_sampled[idx] || run_sampled_) {
         manager().end_callbacks[idx](*this);
       }
     }
+
+    AT_ASSERT(thread_local_func_ == this, name_, ": must be top of stack");
     thread_local_func_ = parent_;
+    initialized_ = false;
   }
 }
 
+at::Tensor record_function_enter(const std::string& name) {
+  if (thread_local_func_) {
+    thread_local_func_->end();
```

Comment:
Oh good catch! Could you:

1. (nit) assert the `thread_local_func_`'s name is indeed "record_function_enter"
2. add a comment explaining this (it took me quite a while to figure out why we need to do this :-))

Reply:
done